### PR TITLE
chore(python): autogenerate docs/index.rst

### PR DESCRIPTION
**Pull request by parthea**
This PR is a first step towards auto generating `docs/index.rst` for *most* python repos. 

The function `detect_versions` in `synthtool/languages/python.py` is identical to the one used by `node` [here](https://github.com/googleapis/synthtool/blob/master/synthtool/languages/node.py#L217) so alternatively, we could consider moving the existing function `detect_versions` outside of `node.py`.

### CODE REVIEW

**Comment by parthea** [comment]
As an example, I ran the version of synthtool in [this PR](https://github.com/googleapis/synthtool/pull/1114) in the `python-memcache` repo and auto generated the `docs/index.rst` file.

File: synthtool/languages/python.py [hunk 1]
```diff
@@ -93,6 +94,36 @@ def _get_sample_readme_metadata(sample_dir: Path) -> dict:
...
+    Returns: a list of the subdirectories; for the example above:
+      ['v1', 'v1alpha', 'v1beta']
+      If specified, the default_version is guaranteed to be listed last.
+      Otherwise, the list is sorted alphabetically.
+    """
+    versions = []
+    if os.path.isdir(path):
+        for directory in os.listdir(path):
+            if os.path.isdir(os.path.join(path, directory)):
+                versions.append(directory)
```
Comment (busunkim96) [inline synthtool/languages/python.py:120]: I believe you could glob for directories that look like they have versions:

```py
>>> import glob
>>> from pathlib import Path
>>> list(Path("google").glob("**/*v[1-9]*"))
[PosixPath('google/cloud/automl_v1'), PosixPath('google/cloud/automl_v1beta1')]
```

File: synthtool/languages/python.py [hunk 2]
```diff
@@ -93,6 +94,36 @@ def _get_sample_readme_metadata(sample_dir: Path) -> dict:
...
+    versions.sort()
+    if default_version is not None:
+        versions = [v for v in versions if v != default_version] + [default_version]
```
Comment (busunkim96) [inline synthtool/languages/python.py:123]: Can `default_version` be put first in the list? I think that's the pattern currently used in the docs. https://googleapis.dev/python/automl/latest/index.html

File: tests/test_python_library.py [hunk 3]
```diff
@@ -126,3 +128,47 @@ def test_split_system_tests():
         with open(templated_files / ".kokoro/presubmit/system-3.8.cfg", "r") as f:
             contents = f.read()
             assert "system-3.8" in contents
+
+
+def test_detect_versions_src():
```
Comment (busunkim96) [inline tests/test_python_library.py:133]: Please update these tests to reflect the layout of the Python libraries.

File: tests/test_python_library.py [hunk 4]
```diff
@@ -126,3 +128,47 @@ def test_split_system_tests():
...
+
+def test_detect_versions_staging():
+    temp_dir = Path(tempfile.mkdtemp())
+    staging_dir = temp_dir / "owl-bot-staging"
+    for v in ("v1", "v2", "v3"):
+        os.makedirs(staging_dir / v)
+
+    versions = python.detect_versions(staging_dir)
+    assert ["v1", "v2", "v3"] == versions
+
```
Comment (busunkim96) [inline tests/test_python_library.py:152]: Is the idea that the versions get detected from `owl-bot-staging`?

File: synthtool/languages/python.py [hunk 5]
```diff
@@ -93,6 +94,36 @@ def _get_sample_readme_metadata(sample_dir: Path) -> dict:
     return sample_metadata
 
 
+def detect_versions(
+    path: str = "./src", default_version: Optional[str] = None
```
[the remark this answers is not in the document]
Comment (parthea) [reply synthtool/languages/python.py]: Fixed in https://github.com/googleapis/synthtool/pull/1114/commits/bdee7a88e47056e692be62de7b1f22ea2aaedc68

File: synthtool/languages/python.py (hunk 1 above)
Comment (parthea) [reply synthtool/languages/python.py]: Fixed in https://github.com/googleapis/synthtool/pull/1114/commits/fb23307ccc88cc41f1dde056aeea6fd1632748fe

File: synthtool/languages/python.py (hunk 2 above)
Comment (parthea) [reply synthtool/languages/python.py]: Done in https://github.com/googleapis/synthtool/pull/1114/commits/89e2bc43056196d5c537d8ac5298c269c0c52eb6

File: tests/test_python_library.py (hunk 3 above)
Comment (parthea) [reply tests/test_python_library.py:133]: Done in https://github.com/googleapis/synthtool/pull/1114/commits/bdee7a88e47056e692be62de7b1f22ea2aaedc68

File: tests/test_python_library.py (hunk 4 above)
Comment (parthea) [reply tests/test_python_library.py]: Correct. I've renamed this test to `test_detect_versions_default_path` now that the default path is `owl-bot-staging`

File: synthtool/languages/python.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (parthea) [reply synthtool/languages/python.py]: My mistake, the default path should be `google/cloud`. I've fixed this in https://github.com/googleapis/synthtool/pull/1114/commits/9a6f82ffac0dc752a664fcaf52f23d84dac8fc8f.

File: tests/test_python_library.py (hunk 3 above)
Comment (parthea) [reply tests/test_python_library.py]: My mistake, the default path should be google/cloud. I've fixed this in 9a6f82f.

File: tests/test_python_library.py (hunk 4 above)
Comment (parthea) [reply tests/test_python_library.py]: My mistake, the default path should be google/cloud. I've fixed this in 9a6f82f.

File: synthtool/gcp/templates/python_library/docs/index.rst [hunk 6]
```diff
@@ -0,0 +1,36 @@
+.. include:: README.rst
+
+.. include:: multiprocessing.rst
+{% if versions|length > 1 %}
+This package includes clients for multiple versions of {{ metadata['repo']['name_pretty'] }}.
+By default, you will get version ``{{ versions | first }}``.
+{% endif %}
+{% for version in versions %}
```
Comment (tswast) [inline synthtool/gcp/templates/python_library/docs/index.rst:8]: Do we have any fallbacks to link to reference documentation if there isn't a version / versions? e.g. the "integration" type repos?

Comment (tswast) [reply synthtool/gcp/templates/python_library/docs/index.rst:8]: Or should we expect to exclude the `index.rst` for these repos?

Comment (parthea) [reply synthtool/gcp/templates/python_library/docs/index.rst:8]: I've added a check in `py_library` [here](https://github.com/googleapis/synthtool/blob/add-docs-index-rst-to-template/synthtool/gcp/common.py#L249) to check that the `versions` kwarg exists and `default_version` is not empty before generating index.rst.

**Comment by parthea** [comment]
@busunkim96 @tswast Please could you take another look?

File: synthtool/gcp/common.py [hunk 7]
```diff
@@ -237,6 +245,13 @@ def py_library(self, **kwargs) -> Path:
         if "samples" not in kwargs:
             self.excludes += ["samples/AUTHORING_GUIDE.md", "samples/CONTRIBUTING.md"]
 
+        # Don't add `docs/index.rst` if `versions` is not provided or `default_version` is empty
+        if (
+            "versions" not in kwargs
+            or not kwargs["metadata"]["repo"]["default_version"]
+        ):
+            self.excludes += ["docs/index.rst"]
```
Comment (tswast) [inline synthtool/gcp/common.py:253]: Wonderful!

**Review by tswast — APPROVED** [review]

**Comment by parthea** [comment]
@chingor13 @busunkim96  Please could you review/approve?

**Review by dandhlee — APPROVED** [review]
Seems like from Sphinx side not much has changed, I'll take that as a good thing. LGTM.

**Review by busunkim96 — APPROVED** [review]

File: synthtool/languages/python.py [hunk 8]
```diff
@@ -93,6 +94,50 @@ def _get_sample_readme_metadata(sample_dir: Path) -> dict:
     return sample_metadata
 
 
+def detect_versions(path: str = "./google/cloud") -> List[str]:
```
Comment (bcoe) [inline synthtool/languages/python.py:97]: How hard would it be for us to pull this into a common file, if the logic is identical?

**Review by bcoe — APPROVED** [review]
Not a blocker for me on landing this, but ideally we'd DRY up the detect version logic.

If you don't opt to do so in this PR, mind creating a tracking issue to do so?

**Review by bcoe — APPROVED** [review]
Thank you for cleaning this up @parthea 👏 , reading through the current `owlbot` logic, I saw that versions is actually loaded in a separate code path:

```python
        # Collect the subdirectories of the src directory.
        src = Path("src")
        versions = [v.name for v in src.iterdir() if v.is_dir()]
        # Reorder the versions so the default version always comes last.
        versions = [v for v in versions if v != default_version] + [default_version]
        logger.info(f"Collected versions ${versions} from ${src}")
```

So I think you've gotten rid of some code that would have otherwise have hung around indefinitely.

**Pull request merged**